File: Data Structures & Algorithms/island-perimeter/submission-0.cpp
class Solution {
public:
    int dfs(vector<vector<int>>& grid, int r, int c, vector<vector<bool>>& visited) {
        if (r >= grid.size() || c >= grid[0].size() || r < 0 || c < 0 || !grid[r][c]) {
            return 1;
        }

        if (grid[r][c] && visited[r][c] == true) return 0;

        visited[r][c] = true;
        int perim = 0;
        perim += dfs(grid, r, c+1, visited) + dfs(grid, r+1, c, visited) + dfs(grid, r, c-1, visited) + dfs(grid, r-1, c, visited); 
        return perim;
    }

    int islandPerimeter(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();

        vector<vector<bool>> visited(rows, vector<bool>(cols, false));

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c]) return dfs(grid, r, c, visited);
            }
        }

        return 0;
    }
};

File: Data Structures & Algorithms/surrounded-regions/submission-0.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        ROWS, COLS = len(board), len(board[0])
        visited, safe = set(), set()
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        def dfs(r, c, visited):
            if (r, c) in visited or r < 0 or c < 0 or r == ROWS or c == COLS or board[r][c] != "O":
                return
            
            visited.add((r, c))
            safe.add((r, c))
            for dr, dc in directions:
                dfs(r+dr, c+dc, visited)
        
        for c in range(COLS):
            if board[0][c] == "O":
                dfs(0, c, visited)

            if board[ROWS-1][c] == "O":
                dfs(ROWS-1, c, visited)
        
        for r in range(ROWS):
            if board[r][0] == "O":
                dfs(r, 0, visited)

            if board[r][COLS-1] == "O":
                dfs(r, COLS-1, visited)

        for r in range(ROWS):
            for c in range(COLS):
                if board[r][c] == "O" and (r, c) not in safe:
                    board[r][c] = "X"

